src/toil/cwl/cwltoil.py

```diff
@@ -193,6 +193,78 @@ class SkipNull:
     """
 
 
+def trim_mounts_op_down(file_or_directory: CWLObjectType) -> None:
+    return
+
+
+def sniff_location(file_or_directory: CWLObjectType) -> Optional[str]:
+    """
+    Return None if we don't have a local path or file URI
+    :param file_or_directory:
+    :return:
+    """
+    # Since we only consider mountable paths, if path is not file URI or bare path, don't consider it
+    path = cast(str, file_or_directory.get('location') or file_or_directory.get('path'))
+    scheme = urlparse(path).scheme
+    if scheme == '' or scheme == 'file':
+        return path
+    else:
+        return None
+
+
+def trim_mounts_op_up(file_or_directory: CWLObjectType, op_down_ret: None, child_results: list[bool]) -> bool:
+    """
+    Nonredundant for something in a directory means its path or location is not within the parent directory or doesn't match its basename
+    Nonredundant for something in a secondary file means its path or location is not adjacent to the primary file or doesn't match its basename
+
+    If on a File:
+    Returns True if anything in secondary files is nonredundant or has nonredundant children to this file, false otherwise
+    If on a Directory:
+    Returns True if anything in top level listing is nonredundant or has nonredundant children, otherwise false.
+    If something in the listing is redundant and all children are redundant, then delete it
+    :param file_or_directory: CWL file or CWL directory type
+    :return: boolean
+    """
+    if file_or_directory.get('location') is None and file_or_directory.get('path') is None:
+        # file or directory is defined by contents or listing respectively, this is not redundant
+        return True
+    own_path = sniff_location(file_or_directory)
+    if own_path is None:
+        return True
+    # basename should be set as we are the implementation
+    own_basename = cast(str, file_or_directory['basename'])
+
+    # If the basename does not match the path, then this is nonredundant
+    if not own_path.endswith("/" + own_basename):
+        return True
```

Comment:
We haven't URL-decoded `own_path` for the case where it's a file: URI, so we will think any filename where there are characters that need to be encoded in file: URIs is nonredundant, even when  it's not.

We might want to make `sniff_location()` handle the conversion to bare, URL-decoded paths, so we can work with only those in here.